### ISSUE TITLE
Allow multiple session in parallel 

### ISSUE THREAD

**Issue by smadbe** [France-ioi/AlgoreaBackend#1022]
## Motivations
Currently, if you log in on the platform a device, and then log in on another device, you get disconnected from the first one. This is not really convenient. We should try to improve this process without comprising the security.

## Subtasks
* When re-doing the authentication process, the other sessions should not be dropped if still valid
* (if not already the case) When a user refresh a token, the former token (used to request the refresh) should have its lifetime reduced (to ~2min?) so that we do not end up with too many valid token for a same user at a same time.
* We should make sure that expired tokens always get flushed at a moment so that the DB does not keep growing (prefer cleaning while we create new tokens instead of requiring a cron job)

**Comment by GeoffreyHuck** [comment]
Observation: The lifetime is currently not reduced for the old token. It has to be added.

### TLDR:
1. What happens if a token has been refreshed, but the old one is used to refresh it again, within the 2min? It may happen if more than one tab is open. **Proposition**: send back the new valid token
2. We need to add a column `sessions`.`old_access_token`, which contains the `access_token` from which we obtained this one during a refresh. This makes it easy to:
  - Find the new access token for the process in point `1.`
  - Remove the old session
3. Someone could open many sessions in parallel without refreshing the token, by cleaning the browser, or programmatically. This can add many rows in `sessions`. **Proposition**: limit the number of parallel sessions to 10.


## The details:

Here's a review on the process of auth & refreshing the token, and what should be modified:

### Current process: when a user is authenticated:
- Add a row in table `sessions` with the new `access_token`
- Update `refresh_tokens` of the user in database

### Current process: when the token is refreshed:
- Back-end calls the Login Service with the `refresh_token` currently in database, and gets a new `access_token` and `refresh_token`
- Deletes all `sessions` of `user_id` with `access_token` != old access token
- We add a row in table `sessions` with the new access token
- Update the `refresh_token` in database

### Current process: notes on `sessions`
- At any point, we have a maximum of 2 entries for a `user_id`: the current `access_token`, and the old one used when the token was refreshed.
- Only the refresh of the token leads to the deletion of entries in `sessions`


### New process with multiple sessions
- `expires_in` is here to simplify the example, the database contains the actual date of expiration.
- **Proposition**: `old_access_token` will be added in the `sessions` table, so we know what rows we need to delete during a refresh token.

Assume we already have 2 entries for a session open on a desktop (A):
```
| access_token | expires_in | old_access_token |
| A1           | 2min       | null             |
| A2           | 1h         | A1               |
```

We log with a mobile (B), we now have:
```
Request access_token: /
| access_token | expires_in | old_access_token |
| A1           | 2min       | null             |
| A2           | 1h         | A1               |
| B1           | 1h         | null             |
```

**What can happen from the desktop (A):**
- The desktop (A) asks for a refresh, in which case we want to
  - Remove `A1`
  - Add `A3`, the new access token for desktop
  - Reduce the lifetime of `A2` to 2min
  - Keep `B1`
```
Request access_token: A2
| access_token | expires_in | old_access_token |
| A1           | 2min       | null             | -- remove
| A2           | 2min       | A1               | -- update expiration
| B1           | 1h         | null             |
| A3           | 1h         | A2               | -- add
```
- **The old access token `A1` is reused and asks for a refresh: !!! THIS CASE IS WEIRD !!!**, but it can happen if we have more than one tab open.
  - **Proposition**: If `A1` is used, we know there is a row with `old_access_token`=`A1` corresponding to `A2`. So we can send back `A2` (if it is not expired, but if it is, then `A1` would have expired already).

**What can happen from the mobile (B):**
- The mobile (B) asks for a refresh
  - Add `B2`, the new access token for mobile
  - Reduce the lifetime of `B1` to 2min
  - Keep `A1` and `A2`
```
Request access_token: B1
| access_token | expires_in | old_access_token |
| A1           | 2min       | null             |
| A2           | 1h         | A1               |
| B1           | 2min       | null             | -- update expiration
| B2           | 1h         | B1               | -- add
```

**Possibility of a new authentication:**
- The user gets authenticated on a third device, or a device gets cleaned (C):
  - Add `C1`
```
Request access_token: /
| access_token | expires_in | old_access_token |
| A1           | 2min       | null             |
| A2           | 1h         | A1               |
| B1           | 1h         | null             |
| C1           | 1h         | null             | -- add
```

### New behavior
When refreshing a session:
- Update `sessions` of the access token used by the request refreshing the session: `expires_at`=`NOW() + 2 minutes`.
- From the row we just updated, if the `old_access_token` `IS NOT NULL`, we remove the row with `access_token`=`old_access_token`. This remove the old session without touching the other parallel sessions.

### New session spamming
It's possible that a client creates many sessions in parallel.

The best way to mitigate this might be to limit the number of sessions. When a new session is created, we make sure the user don't have more than 10 sessions (20 rows in `sessions`), by removing them from the oldest to the newest, ordered by expiration date.

**Comment by smadbe** [comment]
Note 1: Your "New process with multiple sessions" scenario is a bit strange as: if A1 expires in 2min, it means, it has just been renewed... so you are already refreshing twice in a row... anyway, in this case, I do not see why A1 should be removed, its lifetime is very limited anyway.

Note 2: I missed the "tab fork" scenario. Because of that I think we cannot reduce the lifetime of the previous token as I proposed and that's all. Otherwise, it would mean that some tab may have outdated information of the lifetime of the cookie they use with potential strange behavior. Anyway it is not look as the former token was compromised. (and if it is, the user should logout and this should remove the token).

Note 3: You do not document what happens with the log out. Remove only the token ? All user session? (on another device as well). I think on Google accounts, all tabs of my browser disconnect if I disconnect, but not my other devices.

Note 4: Some browsers (chrome in particular since recently) make the non-visible tabs unable to execute a lot of code and follow clock... so we may consider that some non-visible tabs may be unable to refresh their token in time.

Note 5: At the end, I am not really convince by the `old_access_token` column. To me, it looks like something to help linking two "related" tokens if they are not too far away... but only at a distance of 1... but 2, 3, infinite may make sense.. 0 as well... so that looks like just a fix but that does not solve all cases.

Note 6: New session spamming -> good idea. 20 rows in sessions? hum that does not mean 10 sessions from what I understood

**Comment by smadbe** [comment]
Discussion result: Geoffrey needs to check whether cookies of different tabs on the same domain share cookies in "real time".

**Comment by GeoffreyHuck** [comment]
Result of the test : **YES**, the cookie is updated in real time (tested with Chrome & Firefox). It makes sense because the browser stores a cookie only once.


### TLDR: Recap of proposed changes (all reasons are documented below)

- ‌Do not change the expiry date of the old token on refresh
- Add a `session_id (int64)` field in table `sessions`. Set to a random number on successful login. New entries in table `sessions` from a token refresh will have the same `session_id` as the token used for the refresh. Support `null` for existing sessions at the moment of migration.
- `logout` service : remove entries in `sessions` matching `user_id`+`session_id`
- `refreshToken` service: if we try to refresh a token with a valid old token, we return the most recent token without refreshing it. Make sure the `expires_in` in the response is properly set.
- `accessTokenCreate` service: To avoid session spamming: If we open a new session and there is more than 10 sessions in the database, we remove the one with the oldest `issued_at` date (remove entries matching `user_id`+`session_id`.
- Shouldn't we consider returning the same token if its expiry date > SOME_TIME to avoid many refresh one after another?
- Document the reason why we check that `len(accessToken) <= 2000`


### Setting the expiry date of the old token, from which the token has been updated, to 2 minutes

As discussed, the purpose of this is to make sure the requests that were already started with the old token are not rejected.

**Current behavior:** The expiry date is not changed, but we remove the older sessions (we keep the current and the one from which it has been refreshed in database).

`Do we really need to update the expiration of the older token since we remove the older ones anyway?`

### Client-side refresh timer

When the app is open in many tabs, we have as many refresh timers going on. Doesn't this mean that all the tabs will ask for a token refresh at about the same time? And that the token will be refreshed many times in a row?

`Shouldn't we consider returning the same token if it has just been refreshed? If its expiry date > SOME_TIME? in order to avoid many refresh one after another?`

### When an old token is used for a refresh

This could happen when:
* Someone does it programmatically (refresh the token, then use the old one to call refresh again)
* Multiple tabs fire the refresh token **at the same time**. The back-end ensures the refresh is made in a lock, but it still processes all refresh one after another. Depending on the synchronicity of the calls, those can happen:
1. Two or more refresh are made one after another, refreshing the token, then refreshing it again with the new token. This is the same as the above `Client-side refresh timer`. It could be handled in the same way.
2. While the token is being refreshed, a call to refresh is made **with the old token**. The token is refreshed, then refreshed again from the old token.

`Proposition: if we try to refresh the token, using a token that is not the most recent but is still valid, we just send the most recent token (the most recent "sessions"."issued_at") without refresh`

`Warning: we need to make sure the front-end sets the refresh timer with the "expired_in" field returned by the response of the service "accessTokenCreate"`

![image](https://github.com/France-ioi/AlgoreaBackend/assets/5871762/fa9dba4b-c200-46fb-9458-2cc94f123750)

### Uniquely identify sessions (linking access tokens together)

We need a way to differentiate the sessions of a user to:

- know which session to `logout`
- know which older tokens to remove when we `refresh`

The previous proposal of storing the old access token seems overkill (see `Note 5`).

`Proposal: Add a "session_id" field in table "sessions" to identify the session with "user_id"+"session_id":`
* For existing session at the moment of the migration, `session_id` will be `null`.
* When a new session is open (user logs in), `session_id` is set to a random 64bits number.
* When the session is updated, the new session will have the same "session_id" of the old session (`null` for the sessions opened before the migration).`

Just removing the old session at token refresh could make requests that have already been made fail, if they're at the stage where the access token has not yet been validated. It would be easier, but we might have errors.


### How do we handle the refresh token of France-IOI login-service?

**Currently:** we store the refresh token in the table `refresh_token` for the `user_id` (primary key).

**Do we use the same login-service refresh token for all sessions?**

Technically we can. The refresh is being made by the back-end only after having checked that the current access token is valid. 

Alternatively, we could store the corresponding refresh token for each `session_id` by adding a field `session_id` to the table `refresh_token`, thus having a different `refresh_token` for each session of the user. **BUT**, we need to make sure that France-IOI login-service supports multiple sessions for each user with a different refresh token.

`Proposition: leave the system as it is, using a unique refresh token for each user with the login-service.`


### Logout (`Note 3`)

#### Old behavior of Logout

- Removes all entries in `sessions` matching the current `user_id`
- Removes the entry in `refresh_tokens` matching the current `user_id`

#### New behavior of logout

As discussed, `logout` will remove the session from which the logout has been made. This will disconnect the user from the current browser, but not from other browser or devices.

- Removes all entries in `sessions` matching the current `user_id`+`session_id`
- If we choose to have multiple refresh token from the login-service per user, we would need to remove the corresponding entries in table `refresh_tokens` too. Otherwise, we keep the refresh token.

### What happens if a tab is in "sleep", and the token is not refreshed in time? (`Note 4`)

I don't think we can do anything about it. The call to `/auth/token` with an expired token will return an error.

### Session spamming (`Note 6`)

The number of sessions of a user can be determined by counting the number of distinct `user_id`+`session_id` in table `sessions`.

`Proposition: If we have more than 10 open sessions, we drop the one with the oldest "issued_at" date.`

### Something weird in authentication

There is a suspicious line here : https://github.com/France-ioi/AlgoreaBackend/blob/master/app/auth/middleware.go#L96

It is not document what should happen when the access token is more than 2000 characters. If I understand currently, the authentication would return "success", but the user structure would be empty?

`Is this intended? If it is, we should document the reason.`

**Comment by smadbe** [comment]
> Support null for existing sessions at the moment of migration.

I would just generate a random (or just sequential) session instead of having an invalid value just for the migration (that we will forget to clean afterwards)

**Comment by smadbe** [comment]
> ### Setting the expiry date of the old token, from which the token has been updated, to 2 minutes
> 
> As discussed, the purpose of this is to make sure the requests that were already started with the old token are not rejected.
> 
> **Current behavior:** The expiry date is not changed, but we remove the older sessions (we keep the current and the one from which it has been refreshed in database).
> 
> `Do we really need to update the expiration of the older token since we remove the older ones anyway?`

It would prevent having tons of sessions in parallel if a user uses several tabs: each time the app is relauched, a token is re-requested... and each tab requests a new token at the half of its lifetime.
But the problem is that you cannot reduce the lifetime if you cannot inform the other tabs anyway.

**Comment by smadbe** [comment]
> Result of the test : **YES**, the cookie is updated in real time (tested with Chrome & Firefox). It makes sense because the browser stores a cookie only once.
> ### TLDR: Recap of proposed changes (all reasons are documented below)
> 
>     * ‌Do not change the expiry date of the old token on refresh
> 
>     * Add a `session_id (int64)` field in table `sessions`. Set to a random number on successful login. New entries in table `sessions` from a token refresh will have the same `session_id` as the token used for the refresh. Support `null` for existing sessions at the moment of migration.
> 
>     * `logout` service : remove entries in `sessions` matching `user_id`+`session_id`
> 
>     * `refreshToken` service: if we try to refresh a token with a valid old token, we return the most recent token without refreshing it. Make sure the `expires_in` in the response is properly set.
> 
>     * `accessTokenCreate` service: To avoid session spamming: If we open a new session and there is more than 10 sessions in the database, we remove the one with the oldest `issued_at` date (remove entries matching `user_id`+`session_id`.
> 
>     * Shouldn't we consider returning the same token if its expiry date > SOME_TIME to avoid many refresh one after another?
> 
>     * Document the reason why we check that `len(accessToken) <= 2000`
> 
> 
> ### Setting the expiry date of the old token, from which the token has been updated, to 2 minutes
> 
> As discussed, the purpose of this is to make sure the requests that were already started with the old token are not rejected.
> 
> **Current behavior:** The expiry date is not changed, but we remove the older sessions (we keep the current and the one from which it has been refreshed in database).
> 
> `Do we really need to update the expiration of the older token since we remove the older ones anyway?`
> ### Client-side refresh timer
> 
> When the app is open in many tabs, we have as many refresh timers going on. Doesn't this mean that all the tabs will ask for a token refresh at about the same time? And that the token will be refreshed many times in a row?
> 
> `Shouldn't we consider returning the same token if it has just been refreshed? If its expiry date > SOME_TIME? in order to avoid many refresh one after another?`
> ### When an old token is used for a refresh
> 
> This could happen when:
> 
>     * Someone does it programmatically (refresh the token, then use the old one to call refresh again)
> 
>     * Multiple tabs fire the refresh token **at the same time**. The back-end ensures the refresh is made in a lock, but it still processes all refresh one after another. Depending on the synchronicity of the calls, those can happen:
> 
> 
>     1. Two or more refresh are made one after another, refreshing the token, then refreshing it again with the new token. This is the same as the above `Client-side refresh timer`. It could be handled in the same way.
> 
>     2. While the token is being refreshed, a call to refresh is made **with the old token**. The token is refreshed, then refreshed again from the old token.
> 
> 
> `Proposition: if we try to refresh the token, using a token that is not the most recent but is still valid, we just send the most recent token (the most recent "sessions"."issued_at") without refresh`
> 
> `Warning: we need to make sure the front-end sets the refresh timer with the "expired_in" field returned by the response of the service "accessTokenCreate"`
> 
> ![image](https://private-user-images.githubusercontent.com/5871762/305100223-fa9dba4b-c200-46fb-9458-2cc94f123750.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDgwOTQxODcsIm5iZiI6MTcwODA5Mzg4NywicGF0aCI6Ii81ODcxNzYyLzMwNTEwMDIyMy1mYTlkYmE0Yi1jMjAwLTQ2ZmItOTQ1OC0yY2M5NGYxMjM3NTAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMTZUMTQzMTI3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ODNhNWE3ODhhMWQyZTgzOTkyODUyYjI2MzY4M2VkYWUwZmYwMzUzNDYyNDFiMDJmOTJiOTUwZDMzOWZlNTBhNCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.7pfu12NeXCRbRd698LTMQJdseei84YO-dscAyF22R_k)
> ### Uniquely identify sessions (linking access tokens together)
> 
> We need a way to differentiate the sessions of a user to:
> 
>     * know which session to `logout`
> 
>     * know which older tokens to remove when we `refresh`
> 
> 
> The previous proposal of storing the old access token seems overkill (see `Note 5`).
> 
> `Proposal: Add a "session_id" field in table "sessions" to identify the session with "user_id"+"session_id":`
> 
>     * For existing session at the moment of the migration, `session_id` will be `null`.
> 
>     * When a new session is open (user logs in), `session_id` is set to a random 64bits number.
> 
>     * When the session is updated, the new session will have the same "session_id" of the old session (`null` for the sessions opened before the migration).`
> 
> 
> Just removing the old session at token refresh could make requests that have already been made fail, if they're at the stage where the access token has not yet been validated. It would be easier, but we might have errors.
> ### How do we handle the refresh token of France-IOI login-service?
> 
> **Currently:** we store the refresh token in the table `refresh_token` for the `user_id` (primary key).
> 
> **Do we use the same login-service refresh token for all sessions?**
> 
> Technically we can. The refresh is being made by the back-end only after having checked that the current access token is valid.
> 
> Alternatively, we could store the corresponding refresh token for each `session_id` by adding a field `session_id` to the table `refresh_token`, thus having a different `refresh_token` for each session of the user. **BUT**, we need to make sure that France-IOI login-service supports multiple sessions for each user with a different refresh token.
> 
> `Proposition: leave the system as it is, using a unique refresh token for each user with the login-service.`
> ### Logout (`Note 3`)
> #### Old behavior of Logout
> 
>     * Removes all entries in `sessions` matching the current `user_id`
> 
>     * Removes the entry in `refresh_tokens` matching the current `user_id`
> 
> 
> #### New behavior of logout
> 
> As discussed, `logout` will remove the session from which the logout has been made. This will disconnect the user from the current browser, but not from other browser or devices.
> 
>     * Removes all entries in `sessions` matching the current `user_id`+`session_id`
> 
>     * If we choose to have multiple refresh token from the login-service per user, we would need to remove the corresponding entries in table `refresh_tokens` too. Otherwise, we keep the refresh token.
> 
> 
> ### What happens if a tab is in "sleep", and the token is not refreshed in time? (`Note 4`)
> 
> I don't think we can do anything about it. The call to `/auth/token` with an expired token will return an error.
> ### Session spamming (`Note 6`)
> 
> The number of sessions of a user can be determined by counting the number of distinct `user_id`+`session_id` in table `sessions`.
> 
> `Proposition: If we have more than 10 open sessions, we drop the one with the oldest "issued_at" date.`
> ### Something weird in authentication
> 
> There is a suspicious line here : https://github.com/France-ioi/AlgoreaBackend/blob/master/app/auth/middleware.go#L96
> 
> It is not document what should happen when the access token is more than 2000 characters. If I understand currently, the authentication would return "success", but the user structure would be empty?
> 
> `Is this intended? If it is, we should document the reason.`

`authorized` would keep its initial value, so `false` I assume, so it would fall into `if !authorized {...` condition. Why would it return "success"?

**Comment by GeoffreyHuck** [comment]
Here's a summary of our talk on Friday 16/02/2024.

### We chose not to modify existing token

This implies:
- We abandon the idea of reducing the expiry date of a token after it has been updated
- We don't delete non-expired token. This is a new behavior compared to the current one where after an refresh, the older token was deleted (only the last two tokens were kept at a time).

### Special cases for `refresh service`

- If a new token is being refreshed again within 5 minutes, we just send back the new token.
-> To avoid refresh spamming when many tabs are open.

- If we try to refresh a token which is still valid but not the most recent one, we just return the most recent one.
-> Can happen when multiple refresh from multiple tabs happen at the same time, or programatically.

### Aligning the behavior of refresh for temp users

The current behavior when the user is temporary is to delete the session (session=access token in the old database schema) and return a new one. This is different to the behavior of a normal user, where the older session is kept until it expired.

We keep it on normal users to make sure no request made before/during the refresh will fail.

So we need to do the same with temp users.

### New database schema

This is the new schema proposed by Damien:

```
sessions table:
session_id (PK)
user_id (FK)
refresh_token (optional)

access_tokens table:
token 
session_id (FK)
expires_at
issued_at
```

The `issuer`column has been deemed unnecessary, because we know if the user is a temp user just by looking at the user.

Also in the future, if we want a temp user session to be transformed into a logged in user session, it will be an operation on the user.

### Regarding the line ` if len(accessToken) <= 2000`

Damien found it was because the max length of the token in database was 2000. We'll rewrite it to make it fail explicitly, because it currently fails because of a border effect. This way it will not look like a backdoor anymore!

### Updated proposal list of changes
- Migrate to the new database schema
- Update the services to use the new database schema
- `logout service`: remove session and access tokens matching `session_id`
- `refreshToken service`: If we try to refresh a token with a valid old token, we return the most recent token without refreshing it. Make sure `expires_in` is properly set in the response.
- `refreshToken service`: Delete expired tokens after a refresh.
- `refreshToken service`: If a new token is refreshed withing 5min, we just return the same token. Make sure `expires_in` is properly set in the response.
- `refreshToken service`: If the user is a temp user, use the same process as for normal users: keep the old access token until it expires instead of deleting it.
- `accessTokenCreate service`: To avoid session spamming: If we open a new session and there is more than 10 sessions in the database, we remove the oldest.
- Rewrite the code around `len(accessToken) <= 2000` to make it fail explicitly, with a comment on the reason.

### Next actions

1. I need to verify if we can handle multiple `refresh_token` with France-IOI `login-service`. More specifically:
a. Login for session A and get refresh token A
b. Login for session B and get refresh token B
c. Verify I can refresh the session A with the refresh token A, even if it's not the last refresh token sent by the `login-service`

2. Make sure the 5min delay for refreshing to a new token cannot mess up with the formula used on the front-end, were the timer is set to `max(half of token lifeftime, 5min)`

3. Validate whether the correctness of the new database schema, and have a migration strategy

**Comment by GeoffreyHuck** [comment]
```
1. I need to verify if we can handle multiple refresh_token with France-IOI login-service. More specifically:
    a. Login for session A and get refresh token A
    b. Login for session B and get refresh token B
    c. Verify I can refresh the session A with the refresh token A, even if it's not the last refresh token sent by the login-service
```
-> It works

```
2. Make sure the 5min delay for refreshing to a new token cannot mess up with the formula used on the front-end, were the timer is set to max(half of token lifeftime, 5min)
```
-> Okay

```
3. Validate whether the correctness of the new database schema, and have a migration strategy
```
-> Okay. The migration will be as follow, first populating the new `sessions`, then `access_tokens`, then cleaning:
* For each row in the current `refresh_tokens` table [create rows for real users in `sessions`]:
  * Create an entry in the new `sessions` table with `user_id`, `refresh_token` and `session_id` (auto incremented for the migration, it will be set as a random `uint63` for new entries added by the back-end)
* For each row in the current `sessions` table linked to a `temp user` [create rows for temp users in `sessions`]:
  * Create an entry in the new `sessions` table with `user_id`, `refresh_token=NULL` and `session_id` (auto incremented for the migration, it will be set as a random `uint63` for new entries added by the back-end)
* For each row in the current `sessions` table [create rows for both real and temp users in `access_tokens`]:
  * Create an entry in the new`access_tokens` table with the `token`, `expires_at`, `issued_at`, and the `session_id` retrieved with the `user_id` in the new table `sessions`
* Remove the rows in `access_token` where `expires_at < NOW()`
* Remove the `sessions` not linked to any `access_token`

**Comment by GeoffreyHuck** [comment]
- [ ] Migrate to the new database schema
- [ ] Update the services & tests to use the new database schema
- [ ] `logout service`: remove session and access tokens matching `session_id`
- [ ] `refreshToken service`: If we try to refresh a token with a valid old token, we return the most recent token without refreshing it. Make sure `expires_in` is properly set in the response.
- [ ] `refreshToken service`: Delete expired tokens after a refresh.
- [ ] `refreshToken service`: If a new token is refreshed withing 5min, we just return the same token. Make sure `expires_in` is properly set in the response.
- [ ] `refreshToken service`: If the user is a temp user, use the same process as for normal users: keep the old access token until it expires instead of deleting it.
- [ ] `accessTokenCreate service`: To avoid session spamming: If we open a new session and there is more than 10 sessions in the database, we remove the oldest.
- [ ] Make sure we use the same processes for temp users and real users
- [ ] Rewrite the code around `len(accessToken) <= 2000` to make it fail explicitly, with a comment on the reason.
